### DISCUSSION TITLE
Linux Icon solution idea

### DISCUSSION THREAD

**Discussion by Avasam** [pyinstaller/pyinstaller#7599] [Features & Development]
A while ago I ported one of my applications that are using PyInstaller over to Linux. It works great, but I'm still bothered that it doesn't have an icon.

I know that on Linux, single-file executables don't have icon metadata. And my understanding is that normally applications add their icons to the desktop environment as part of the install process.

So I've been thinking, what if PyInstaller installs and sets an icon as part of the bootstrapping process it does? After executing the application once, the executable will have an icon!

The idea would be to extract the icon to a standard location (see
https://martin.hoppenheit.info/blog/2016/where-to-put-application-icons-on-linux
https://specifications.freedesktop.org/icon-theme-spec/icon-theme-spec-0.13.html#install_icons
for potential locations, I'm no Linux pro myself)
Then run some variant of `gio set ./my_frozen_app metadata::custom-icon file://./path/to/icons/locations/PyInstaller/my_frozen_app_icon`

Some thoughts may have to be put into making sure to avoid filename conflicts.

---

Edit:
For posterity, on Linux I found it's so much better to just build in onedir mode and bundle as an AppImage https://appimage.github.io/appimagetool/
It's well supported, and can:
- come with an icon (no separate .desktop file needed, you provide your own static one at build time)
- self-updating information (using 3rd party user-facing appimage updater tools)
- Can be automatically registered (with same tools, personally I've been using GearLever)
- embed metadata (so when it's registered as an app for your desktop/search menu, it's easy to find)
- Faster self-extracting startup (which also means no need for splash screen, so no splash screen transparency issue)

My setup for one of my apps: https://github.com/Toufool/AutoSplit/blob/v2.3.3/scripts/build.ps1

**Comment by bwoodsend** [comment]
The standard icons directories (`/usr/share/icons/hicolor/$size/apps/`) are owned by root so they can't be written to there.

>  Then run some variant of gio set ./my_frozen_app metadata::custom-icon file://./path/to/icons/locations/PyInstaller/my_frozen_app_icon

Does this actually work for you? My desktop environment is ignoring it. I see that there's also a `standard::symbolic-icon` and a `standard::icon` set by default however neither are writable.

**Reply by bwoodsend** [reply]
> The standard icons directories (/usr/share/icons/hicolor/$size/apps/) are owned by root so they can't be written to there.

Actually, there is a `~/.local/share/icons` too.

**Comment by rokm** [comment]
> I know that on Linux, single-file executables don't have icon metadata. And my understanding is that normally applications add their icons to the desktop environment as part of the install process.

Normally GUI applications provide a .desktop file so they can be discovered and launched from the desktop environment's launcher. With the added benefit that they can specify localized program description, additional arguments, etc. 

And non-GUI applications are intended to be launched from command-line, so they do not need an icon.

> So I've been thinking, what if PyInstaller installs and sets an icon as part of the bootstrapping process it does? After executing the application once, the executable will have an icon!

As far as I'm concerned, neither the bootloader nor bootstrap scripts are the place for this sort of magic.

But nothing prevents you from implementing the logic in your application code, if you really want your application to install the icon for itself and modify the gio metadata.

**Reply by Avasam** [reply]
> But nothing prevents you from implementing the logic in your application code, if you really want your application to install the icon for itself and modify the gio metadata.

Indeed, that's exactly what I've been playing around with. Decided to share the idea to get some feedback (since I'm new to linux development).
Also thought that if a not-too-hacky solution can be found, PyInstaller could offer a standardized way to solve this issue for everyone, and essentially bring its linux features closer to parity with other OSes.

> As far as I'm concerned, neither the bootloader nor bootstrap scripts are the place for this sort of magic.

Alright fair. What about a utility method from PyInstaller's API? There's precedent with the splash screen methods available only from a frozen app after launch.

**Reply by bwoodsend** [reply]
The splash screen stuff is only part of PyInstaller's API because it requires special integration with the bootloader. What you're suggesting would only need a bit of Python code near the start of your script. If you went down the route of generating a .desktop file, it wouldn't even be specific to frozen applications - you could set the `Exec:` field to `f"{sys.executable}" "{sys.argv[0]}"` then you'd have an icon for a non frozen application.

If it were part of PyInstaller then all the questions like what to do if the icons are already there or what sizes to use (desktop's typically prefer icons to be made available in a range of sizes to minimize the amount of scaling it has to do), should the install be triggered always or by a CLI flag or environment variable, which everyone will have a different opinion on would all become our problem to implement and support and test and maintain which we don't want to do.

**Comment by Avasam** [comment]
For posterity, on Linux I found it's so much better to just build in onedir mode and bundle as an AppImage https://appimage.github.io/appimagetool/
It's well supported, and can:
- come with an icon (no separate .desktop file needed, you provide your own static one at build time)
- self-updating information (using 3rd party user-facing appimage updater tools)
- Can be automatically registered (with same tools, personally I've been using GearLever)
- embed metadata (so when it's registered as an app for your desktop/search menu, it's easy to find)
- Faster self-extracting startup (which also means no need for splash screen, so no splash screen transparency issue)

My setup for one of my apps: https://github.com/Toufool/AutoSplit/blob/v2.3.3/scripts/build.ps1
